### PR TITLE
rpc: Add test-only RPC getaddrmaninfo for new/tried table address count

### PR DESCRIPTION
**Pull request by stratospher**
implements https://github.com/bitcoin/bitcoin/issues/26907. split off from #26988 to keep RPC, CLI discussions separate.

This PR introduces a new RPC `getaddrmaninfo`which returns the count of addresses in the new/tried table of a node's addrman broken down by network type. This would be useful for users who want to see the distribution of addresses from different networks across new/tried table in the addrman.

```jsx
$ getaddrmaninfo

Result:
{                   (json object) json object with network type as keys
  "network" : {     (json object) The network (ipv4, ipv6, onion, i2p, cjdns)
    "new" : n,      (numeric) number of addresses in new table
    "tried" : n,    (numeric) number of addresses in tried table
    "total" : n     (numeric) total number of addresses in both new/tried tables from a network
  },
  ...
}
```

### additional context from [original PR](https://github.com/bitcoin/bitcoin/pull/26988)

1. network coverage tests were skipped because there’s a small chance that addresses from different networks could hash to the same bucket and cause count of different network addresses in the tests to fail. see https://github.com/bitcoin/bitcoin/pull/26988#discussion_r1137596851.
2. #26988 uses this RPC in -addrinfo CLI. Slight preference for keeping the RPC hidden since this info will mostly be useful to only super users. see https://github.com/bitcoin/bitcoin/pull/26988#discussion_r1173964808.

### CODE REVIEW

**Review by brunoerg — COMMENTED** [review]
Concept ACK

**Review by brunoerg — APPROVED** [review]
tACK c505611f08a850acb97dea9cc03b36aa468929ca

just tested it with my "clearnet only" node, results:
```sh
➜  bitcoin-core-dev git:(27511-stratospher) ✗ ./src/bitcoin-cli getaddrmaninfo
{
  "ipv4": {
    "new": 28765,
    "tried": 277,
    "total": 29042
  },
  "ipv6": {
    "new": 7245,
    "tried": 67,
    "total": 7312
  },
  "onion": {
    "new": 0,
    "tried": 0,
    "total": 0
  },
  "i2p": {
    "new": 0,
    "tried": 0,
    "total": 0
  },
  "cjdns": {
    "new": 0,
    "tried": 0,
    "total": 0
  }
}
```

File: test/functional/rpc_net.py [hunk 1]
```diff
@@ -328,6 +329,19 @@ def test_addpeeraddress(self):
...
+
+        self.log.debug("Test that getaddrmaninfo is a hidden RPC")
+        # It is hidden from general help, but its detailed help may be called directly.
+        assert "getaddrmaninfo" not in node.help()
+        assert "getaddrmaninfo" in node.help("getaddrmaninfo")
+
+        # current count of ipv4 addresses in addrman is {'new':1, 'tried':1}
+        res = node.getaddrmaninfo()
+        assert_equal(res["ipv4"]["new"], 1)
+        assert_equal(res["ipv4"]["tried"], 1)
```
Comment (kevkevinpal) [inline test/functional/rpc_net.py:344]: Do you think we can add total to this test aswell?

**Review by mzumsande — COMMENTED** [review]
Code Review ACK c505611f08a850acb97dea9cc03b36aa468929ca, I would use this rpc a lot in my work.

I also think that a `totals` field would be useful. 

I don't think that a split-up with respect to "terrible" would be necessary, because in my opinion "terrible" is an internal property used only in address relay / addrman collisions that is not so critical to expose because it isn't used for the main purpose of addrman: selecting peers for outbound connections

**Comment by mzumsande** [comment]
> It's used in RPC getnodeaddresses and CLI -addrinfo to return only non-IsTerrible peers to users. (I meant pre and post that filtering in the suggestion above).

in my opinion unfortunately. I think this was mostly coincidental, because #13152 chose to reuse an addrman function used only for GETADDR answers before. Then #21595 used that output without any mentioning of that restriction, and the doc was only adjusted a year later in #24370. I think that ideally we never should have exposed the terrible filter in the first place.

**Comment by fanquake** [comment]
> I think that ideally we never should have exposed the terrible filter in the first place.

@mzumsande do you want to propose some changes to remove its exposure?

File: src/rpc/net.cpp [hunk 2]
```diff
@@ -966,6 +966,48 @@ static RPCHelpMan addpeeraddress()
...
+
+                          UniValue ret(UniValue::VOBJ);
+                          for (int n = 0; n < NET_MAX; ++n) {
+                              enum Network network = static_cast<enum Network>(n);
+                              if (network == NET_UNROUTABLE || network == NET_INTERNAL) continue;
+                              UniValue obj(UniValue::VOBJ);
+                              obj.pushKV("new", node.addrman->Size(network, true));
+                              obj.pushKV("tried", node.addrman->Size(network, false));
+                              obj.pushKV("total", node.addrman->Size(network));
+                              ret.pushKV(GetNetworkName(network), obj);
```
[the remark this answers is not in the document]
Comment (stratospher) [reply src/rpc/net.cpp:1005]: done. liked the idea.

File: test/functional/rpc_net.py (hunk 1 above)
Comment (stratospher) [reply test/functional/rpc_net.py:344]: yes! done now.

**Review by mzumsande — COMMENTED** [review]
Tested ACK 7d7ee5e0f4254495349ea68f3d0127c0c336fd12

**Comment by fanquake** [comment]
> no, I don't want to spend time on discussions on whether some change breaks user space or not.

@mzumsande fair enough. I don't quite understand what "user space" means, when it has been used in comments here (and in other PRs), as we are not the Linux Kernel, but I assume it's a somewhat less direct way of saying backwards compatibility/continuity?

In any case, as far as I'm aware, we offer no special "backwards compatiblity", or other, guarantees for our CLI tools, and I don't really think we should. Doing so would put us in an awkward position, of having our RPCs backwards compatibility requirements, extended to those tools (consumers, which just happen to exist in this codebase).

We don't want to be in position where we can't change an RPC, to stop it doing something possibly incorrect/illogical, just because a CLI tool happens to call that RPC. These tools shouldn't get any special treatment, and, as far as I'm concerned, should just adapt to any RPC changes we might make, like every other consumer (internal or external) of that RPC.

**Comment by ryanofsky** [comment]
The code change here seems very simple, but since this PR does add a new RPC it would be useful to have more concept ACKs to know whether this is appropriate to merge. I noticed a number of contributors added github reactions to some comments here, so perhaps more of the people following the discussion could add concept acks or mention any reservations

**Review by ryanofsky — APPROVED** [review]
Code review ACK 7d7ee5e0f4254495349ea68f3d0127c0c336fd12

File: test/functional/rpc_net.py (hunk 1 above)
Comment (Jones098) [reply test/functional/rpc_net.py:344]: Thanks all

File: src/rpc/net.cpp [hunk 3]
```diff
@@ -967,6 +967,55 @@ static RPCHelpMan addpeeraddress()
...
+                      "addresses in the `new` and `tried` tables and their sum for all networks.\n"
+                      "This RPC is for testing only.\n",
+                      {},
+                      RPCResult{
+                              RPCResult::Type::OBJ_DYN, "", "json object with network type as keys",
+                              {
+                                      {RPCResult::Type::OBJ, "network", "the network (" + Join(GetNetworkNames(), ", ") + ")",
+                                       {
+                                               {RPCResult::Type::NUM, "new", "number of addresses in the new table, which represent potential peers the node has discovered but hasn't yet successfully connected to."},
+                                               {RPCResult::Type::NUM, "tried", "number of addresses in the tried table, which represent peers with whom the node has successfully connected to in the past."},
```
Comment (mzumsande) [inline src/rpc/net.cpp:983]: grammar nit: would drop "with whom"

**Review by mzumsande — COMMENTED** [review]
ACK 6d7fd9fb9b38be75b2bce643a99cc584837a902b

**Comment by mzumsande** [comment]
ACK [3803efb](https://github.com/bitcoin/bitcoin/commit/3803efb6f6336d50ba0a8367b6f3fc8c66f60664), only change was to improve the RPC doc.

**Comment by mzumsande** [comment]
ACK e4f9b4097e8fd2e3aeb9bf0683cd0d9c1199319d (only rebase)

**Review by 0xB10C — COMMENTED** [review]
Reviewed the changes and manually tested the RPC. Looks good besides maybe renaming `all networks` to `all_networks`.

**Comment by 0xB10C** [comment]
It came up in a discussion that a follow-up to this PR could add a list of addresses (with source, time added, bucket, position, network, ...) to this RPC with a `verbose` flag.

File: test/functional/rpc_net.py [hunk 4]
```diff
@@ -360,6 +361,27 @@ def test_sendmsgtopeer(self):
...
+    def test_getaddrmaninfo(self):
+        self.log.info("Test getaddrmaninfo")
+        node = self.nodes[1]
+
+        self.log.debug("Test that getaddrmaninfo is a hidden RPC")
+        # It is hidden from general help, but its detailed help may be called directly.
+        assert "getaddrmaninfo" not in node.help()
+        assert "getaddrmaninfo" in node.help("getaddrmaninfo")
+
+        # current count of ipv4 addresses in addrman is {'new':1, 'tried':1}
```
Comment (willcl-ark) [inline test/functional/rpc_net.py:373]: nit: if you re-touch could add a debug log line what part of the new functionality we are testing as on L368

**Review by willcl-ark — APPROVED** [review]
ACK e4f9b4097e

I would like to be a user of this data. Like @0xB10C I would be interested in seeing a followup to this which exposed even more granular details, to permit monitoring of the node's addrman without the need for tracepoints or logging.

Happy to re-ACK if the JSON key gets updated.

**Review by brunoerg — APPROVED** [review]
ACK e4f9b4097e8fd2e3aeb9bf0683cd0d9c1199319d

Just tested it again and got expected result. I agree on JSON key getting updated and can reACK it as soon as it gets updated.

**Review by brunoerg — APPROVED** [review]
reACK 0bff92a3b34fea764efdfc474e4d69bb7da0882e

edit: needs to update the functional tests

File: test/functional/rpc_net.py [hunk 5]
```diff
@@ -360,6 +361,28 @@ def test_sendmsgtopeer(self):
...
+        self.log.info("Test that count of addresses in addrman match expected values")
+        res = node.getaddrmaninfo()
+        assert_equal(res["ipv4"]["new"], 1)
+        assert_equal(res["ipv4"]["tried"], 1)
+        assert_equal(res["ipv4"]["total"], 2)
+        assert_equal(res["all networks"]["new"], 1)
```
Comment (brunoerg) [inline test/functional/rpc_net.py:379]:
```suggestion
        assert_equal(res["all_networks"]["new"], 1)
```

**Comment by 0xB10C** [comment]
ACK 28bac81a346c0b68273fa73af924f7096cb3f41d

File: src/rpc/net.cpp [hunk 6]
```diff
@@ -1016,6 +1016,55 @@ static RPCHelpMan sendmsgtopeer()
...
+                              obj.pushKV("new", node.addrman->Size(network, true));
+                              obj.pushKV("tried", node.addrman->Size(network, false));
+                              obj.pushKV("total", node.addrman->Size(network));
+                              ret.pushKV(GetNetworkName(network), obj);
+                          }
+                          UniValue obj(UniValue::VOBJ);
+                          obj.pushKV("new", node.addrman->Size(std::nullopt, true));
+                          obj.pushKV("tried", node.addrman->Size(std::nullopt, false));
+                          obj.pushKV("total", node.addrman->Size());
+                          ret.pushKV("all networks", obj);
```
[the remark this answers is not in the document]
Comment (stratospher) [reply src/rpc/net.cpp]: true! done.

File: test/functional/rpc_net.py (hunk 4 above)
Comment (stratospher) [reply test/functional/rpc_net.py:373]: done.

File: test/functional/rpc_net.py (hunk 5 above)
Comment (stratospher) [reply test/functional/rpc_net.py]: done.

**Review by brunoerg — APPROVED** [review]
reACK 28bac81a346c0b68273fa73af924f7096cb3f41d

CI failure seems unrelated.

File: src/rpc/net.cpp [hunk 7]
```diff
@@ -1016,6 +1016,55 @@ static RPCHelpMan sendmsgtopeer()
...
+                      RPCExamples{
+                              HelpExampleCli("getaddrmaninfo", "")
+                              + HelpExampleRpc("getaddrmaninfo", "")
+                      },
+                      [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
+                      {
+                          NodeContext& node = EnsureAnyNodeContext(request.context);
+                          if (!node.addrman) {
+                              throw JSONRPCError(RPC_CLIENT_P2P_DISABLED, "Error: Address manager functionality missing or disabled");
+                          }
```
Comment (theStack) [inline src/rpc/net.cpp:1046]: nit: to deduplicate code, could introduce a helper function `EnsureAddrman` in rpc/server_util.cpp (similar to the already existing `EnsureConnman` and `EnsurePeerman`) and use that here and in the `addpeeraddress` RPC.

**Review by theStack — APPROVED** [review]
Code-review ACK 28bac81a346c0b68273fa73af924f7096cb3f41d

**Comment by willcl-ark** [comment]
reACK 28bac81a346c0b68273fa73af924f7096cb3f41d

**Comment by achow101** [comment]
ACK 28bac81a346c0b68273fa73af924f7096cb3f41d

File: src/rpc/net.cpp [hunk 8]
```diff
@@ -1016,6 +1016,55 @@ static RPCHelpMan sendmsgtopeer()
...
+{
+    return RPCHelpMan{"getaddrmaninfo",
+                      "\nProvides information about the node's address manager by returning the number of "
+                      "addresses in the `new` and `tried` tables and their sum for all networks.\n"
+                      "This RPC is for testing only.\n",
+                      {},
+                      RPCResult{
+                              RPCResult::Type::OBJ_DYN, "", "json object with network type as keys",
+                              {
+                                      {RPCResult::Type::OBJ, "network", "the network (" + Join(GetNetworkNames(), ", ") + ")",
```
Comment (0xB10C) [inline src/rpc/net.cpp:1029]: nit: This is missing the `all_networks` key in the RPC help.

Can add a fix for this in https://github.com/bitcoin/bitcoin/pull/28523 as this touches the line.

Comment (0xB10C) [reply src/rpc/net.cpp:1029]: included in #28523

Comment (stratospher) [reply src/rpc/net.cpp:1029]: done in #28565

File: src/rpc/net.cpp (hunk 7 above)
Comment (stratospher) [reply src/rpc/net.cpp:1046]: done in #28565.

**Pull request merged**